packages/xo-server/src/index.js

```diff
@@ -140,6 +148,54 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  express.get('/signin/otp', (req, res, next) => {
+    const { user } = req.session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    res.send(
+      signInPage({
+        error: req.flash('error')[0],
+        otp: true,
+        strategies,
+      })
+    )
+  })
+
+  express.post('/signin/otp', (req, res, next) => {
+    const { body, session } = req
+    const { user } = session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    const isValid = authenticator.check(body.otp, user?.preferences?.otp)
+    if (isValid) {
+      setToken(req, res, next)
+    } else {
+      req.flash('error', 'Invalid code')
+      return res.redirect('/signin/otp')
+    }
+  })
+
+  const setToken = async (req, res, next) => {
+    const { user, persistent } = req.session
+
+    // The cookie will be set in via the next request because some
+    // browsers do not save cookies on redirect.
+    req.flash(
+      'token',
+      (await xo.createAuthenticationToken({ userId: user.id })).id
+    )
+    // The session is only persistent for internal provider and if 'Remember me' box is checked
+    req.flash('session-is-persistent', persistent)
```

Comment:
IMHO we should create the `token` cookie now instead of using flash values and then creating it.